crates/ruff_python_parser/src/parser/mod.rs

```diff
@@ -742,23 +814,78 @@ impl RecoveryContextKind {
                         | TokenKind::Lpar
                 )
             }
-            RecoveryContextKind::ImportNames => {
-                matches!(p.current_token_kind(), TokenKind::Rpar | TokenKind::Newline)
+            // The names of an import statement cannot be parenthesized, so it
```

Comment:
I might update the terminator tokens when I write some thorough test cases for invalid programs for better error recovery. But, for now, it's only limited to all valid programs.